e2e testing for PreferSameZone/PreferSameNode

danwinship opened the pull request:
#### What this PR does / why we need it:
e2e tests for PreferSameZone/PreferSameNode

This also extends the existing PreferClose test to test the case where the client and server are in the same zone but not on the same node.

This required a lot of refactoring to avoid just totally duplicating everything multiple times...

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
```docs
- [KEP]: https://github.com/kubernetes/enhancements/issues/3015
```

/kind cleanup
/sig network
/assign @aojea @gauravkghildiyal

danwinship committed:
Add a tests for PreferSameNode

danwinship committed:
Remove a level of indentation in the TrafficDistribution e2e

(No other changes to the code)

danwinship committed:
Streamling the node/zone-picking logic

danwinship committed:
TrafficDistribution e2e cleanups

Remove endpointSlicesHaveSameZoneHints check. We are testing that
connections end up at the right endpoints. We don't need to validate
_why_ they go to the right endpoints, which is already tested by other
tests anyway. (Also, validating the hints becomes more complicated in
the same-node case, where there may or may not also be same-zone hints
depending on cluster configuration.)

Remove DeferCleanup calls; we don't need to delete anything manually
because namespaced resources will automatically be deleted when the
test case's namespace is deleted.

Remove a setting of pod.NodeName that was redundant with
e2epod.SetNodeSelection().

danwinship committed:
Add "multi-node" TrafficDistribution test

The existing TrafficDistribution test didn't really distinguish "same
zone" from "same node". Add another test that makes sure there are at
least 2 nodes in each zone so it can do that.

(Keep the original test as well to avoid losing coverage in CI systems
with single-schedulable-node-per-zone clusters.)

danwinship committed:
Add test for `trafficDistribution: PreferSameZone`

(Identical to the PreferClose test.)

aojea commented:
/test pull-kubernetes-e2e-kind-multizone

danwinship commented:
/hold
to make sure we get all the required test coverage

danwinship commented:
/test pull-kubernetes-e2e-kind-multizone

danwinship committed:
Remove TopologyHints and TrafficDistribution feature flags

The features are always enabled, so the tests don't need to be
conditional.

danwinship commented:
/test pull-kubernetes-e2e-kind-multizone

aojea commented on test/e2e/network/topology_hints.go:
```diff
@@ -41,7 +40,7 @@ import (
 	admissionapi "k8s.io/pod-security-admission/api"
 )
 
-var _ = common.SIGDescribe(feature.TopologyHints, func() {
```
`Feature` in the e2e framework terminology indicates `this test needs to enable something to work, does not run in a normal cluster`

aojea replied:
this test requires a cluster to have zones so we need to add the Feature tag or it will be picked by any normal test

aojea commented on test/e2e/network/traffic_distribution.go:
```diff
@@ -118,49 +118,76 @@ var _ = common.SIGDescribe("Traffic Distribution", func() {
+				break
 			}
 		}
+		if len(nodeForZone) < 3 {
```
wasn't 2 nodes per zone in 3 zones?

danwinship commented:
/test pull-kubernetes-e2e-kind-multizone

gauravkghildiyal commented:
/retest

gauravkghildiyal commented:
Seems like the new tests are not currently running since the cluster doesn't have the feature gate enabled (likely needs something like https://github.com/kubernetes/test-infra/pull/32190) Confirming whether the tests are working correctly in a local kind cluster? @danwinship 

Thanks for the refactor, the test itself LGTM.

danwinship commented:
hm, I thought it was going to get tested in pull-kubernetes-e2e-gce-cos-alpha-features...

well, e2e-kind-multizone at least shows that the refactored old PreferClose test and the new PreferClose test are both passing...

gauravkghildiyal commented:
> hm, I thought it was going to get tested in pull-kubernetes-e2e-gce-cos-alpha-features...

Ahh yeah maybe some of the "multi-node" or "multi-zone" requirements make it skip.

> well, e2e-kind-multizone at least shows that the refactored old PreferClose test and the new PreferClose test are both passing...

Yep true. Thats good. 

/lgtm
/hold Since we are close to freeze, feel free to unhold if we've tested all locally.

danwinship commented:
/test pull-kubernetes-e2e-kind-multizone

danwinship commented:
So it seems that

- e2e-kind-multizone `--skip`s all `[Alpha]` features, independently of what feature gates you have enabled. Given the way `--focus` and `--skip` work, there's no easy way to get it to run these tests (though maybe in the future if we move fully to selecting tests by labels rather than regexes, this will be possible).
- e2e-gce-cos-alpha-features enables all feature gates, but only `--focus`es specific `Feature:` tags. Given that the tagging around features and gates is currently in flux, I imagine this is something that will be improved and made more automatic in the 1.34 cycle...

Anyway, pushed a debug commit to make `PreferSameTrafficDistribution` a GA feature...

gauravkghildiyal commented:
Great! The tests for this PR passed in [pull-kubernetes-e2e-kind-multizone](https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/130945/pull-kubernetes-e2e-kind-multizone/1902900694285815808)

danwinship commented:
/hold cancel

danwinship commented:
> Great! The tests for this PR passed in [pull-kubernetes-e2e-kind-multizone](https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/130945/pull-kubernetes-e2e-kind-multizone/1902900694285815808)

but an SCTP NetworkPolicy test failed because @thockin is psychically taunting me!

gauravkghildiyal commented:
/lgtm

aojea commented on test/e2e/network/traffic_distribution.go:
```diff
@@ -41,17 +41,17 @@ import (
 )
 
-var _ = common.SIGDescribe(feature.TrafficDistribution, framework.WithFeatureGate(features.ServiceTrafficDistribution), func() {
+var _ = common.SIGDescribe("Traffic Distribution", func() {
```
sa in https://github.com/kubernetes/kubernetes/pull/130945/files#r2007167157 , a test that can not be run in a default cluster because some special requirements must use the `Feature` label , we can add `Feature:Multizone` for this and explain that

aojea replied:
```diff
diff --git a/test/e2e/feature/feature.go b/test/e2e/feature/feature.go
index 077f73e6db2..754f81957ab 100644
--- a/test/e2e/feature/feature.go
+++ b/test/e2e/feature/feature.go
@@ -277,6 +277,10 @@ var (
        // Marks tests that require a cloud provider that implements LoadBalancer Services
        LoadBalancer = framework.WithFeature(framework.ValidFeatures.Add("LoadBalancer"))
 
+       // Owner: sig-network
+       // Marks tests that require nodes to belong to different zones
+       MultiZone = framework.WithFeature(framework.ValidFeatures.Add("MultiZone"))
+
        // Owner: sig-storage
        LSCIQuotaMonitoring = framework.WithFeature(framework.ValidFeatures.Add("LSCIQuotaMonitoring"))
```

aojea replied:
```suggestion
var _ = common.SIGDescribe("Traffic Distribution", feature.MultiZone, func() {
```

danwinship replied:
Currently it's pretty consistent across e2e.test that we _don't_ use a feature for this, and just use, eg, `e2eskipper.SkipUnlessMultizone`.

But this is something we can worry about separately from this PR.

aojea replied:
agree

aojea commented on test/e2e/feature/feature.go:
```diff
@@ -512,14 +512,6 @@ var (
 	// - ci-kubernetes-node-e2e-cri-proxy-serial
 	CriProxy = framework.WithFeature(framework.ValidFeatures.Add("CriProxy"))
 
-	// Owner: sig-network
-	// Marks tests that require a cluster with Topology Hints enabled.
-	TopologyHints = framework.WithFeature(framework.ValidFeatures.Add("Topology Hints"))
-
-	// Owner: sig-network
-	// Marks tests that require a cluster with Traffic Distribution enabled.
-	TrafficDistribution = framework.WithFeature(framework.ValidFeatures.Add("Traffic Distribution"))
-
```
self note, both are GA locked by default, this is correct

aojea commented on test/e2e/feature/feature.go:
```diff
@@ -512,14 +512,6 @@ var (
 	// - ci-kubernetes-node-e2e-cri-proxy-serial
 	CriProxy = framework.WithFeature(framework.ValidFeatures.Add("CriProxy"))
 
```
```suggestion
       // Owner: sig-network
       // Marks tests that require nodes to belong to different zones
       MultiZone = framework.WithFeature(framework.ValidFeatures.Add("MultiZone"))
```

aojea commented on test/e2e/network/topology_hints.go:
```diff
@@ -41,7 +40,7 @@ import (
 )
 
-var _ = common.SIGDescribe(feature.TopologyHints, func() {
+var _ = common.SIGDescribe("Topology Hints", func() {
```
```suggestion
var _ = common.SIGDescribe("Topology Hints", feature. MultiZone, func() {
```

aojea commented on test/e2e/network/traffic_distribution.go:
```diff
@@ -41,17 +41,17 @@ import (
 
 	ginkgo.BeforeEach(func(ctx context.Context) {
 		c = f.ClientSet
-		e2eskipper.SkipUnlessAtLeastNZones(ctx, c, 3)
```
better to skip in BeforeEach , just use that checks if there is at least one zone
```
		e2eskipper.SkipUnlessMultizone(ctx, c)
```

danwinship replied:
The PreferSameNode test doesn't require multi-zone though. That's why I moved it. I should probably have mentioned that in the commit message.

aojea commented on test/e2e/network/traffic_distribution.go:
```diff
@@ -119,164 +138,323 @@ var _ = common.SIGDescribe(feature.TrafficDistribution, framework.WithFeatureGat
 	}
 
+	// Data structures for representing client and endpoint pods
+	type endpoint struct {
+		node *v1.Node
+		pod  *v1.Pod
```
endpoints is overloaded and in this test more confusing, what about `server`

danwinship replied:
OK. (I originally had "server" but I kept wanting to refer to them as endpoints in comments/logs, so I changed it.)

aojea commented on test/e2e/network/traffic_distribution.go:
```diff
@@ -119,164 +138,323 @@ var _ = common.SIGDescribe(feature.TrafficDistribution, framework.WithFeatureGat
+			if zone != "" {
+				nodesHaveZones = true
+			}
+			pod := e2epod.NewAgnhostPod(f.Namespace.Name, fmt.Sprintf("endpoint-%d-%s", i, node), nil, nil, nil, "serve-hostname")
+			ginkgo.By(fmt.Sprintf("creating a server pod %q on node %q in zone %q", pod.Name, node, zone))
```
this is what I think we should call this `servers`, endpoint is the representation but this is the realization

aojea commented on test/e2e/network/traffic_distribution.go:
```diff
@@ -119,164 +138,323 @@ var _ = common.SIGDescribe(feature.TrafficDistribution, framework.WithFeatureGat
+		doMultiNodeSameZoneTrafficDistributionTest(ctx, v1.ServiceTrafficDistributionPreferSameZone)
+	})
 
+	framework.It("should route traffic to an endpoint on the same node when using PreferSameNode", framework.WithFeatureGate(features.PreferSameTrafficDistribution), func(ctx context.Context) {
```
I think that the most important part of this feature is
- prefer same node if available
- failover when not available
- when same node is available again prefer same node

This test must guarantee that implementations meet this behavior and is a serious candidate to conformance.
I would not worry about duplicating code if you feel is a stretch to reuse the helper, but I think is important we ensure in this test these 3 things are correct

aojea replied:
nevermind, I got it now, you are testing that with a third client,

aojea replied:
I'm going to delete this comment, I should have read the test entirely first, sorry

aojea commented on test/e2e/network/traffic_distribution.go:
```diff
@@ -119,164 +138,323 @@ var _ = common.SIGDescribe(feature.TrafficDistribution, framework.WithFeatureGat
+				break
+			}
+		}
+		if len(nodes) < 3 {
+			e2eskipper.Skipf("have %d schedulable nodes, need at least 3", len(nodes))
+		}
```
it will be great if we can make this test for just two nodes so it runs in conformance and most of the CI jobs, can we relax this to two nodes and remove the second client with only one endpoint?

danwinship replied:
Hm... we should document the different e2e environments better. I had thought 3 nodes was typical...

Anyway, yes, I can squash this down to just 2 nodes.

aojea replied:
most CsI using kind use 2 nodes that is the minimum for conformance, and kind is used on most project because it runs in github actions conformance clusters ...

danwinship commented on test/e2e/network/traffic_distribution.go:
```diff
@@ -119,164 +138,323 @@ var _ = common.SIGDescribe(feature.TrafficDistribution, framework.WithFeatureGat
+		servingPodLabels := map[string]string{"app": f.UniqueName}
+		for i, ep := range endpoints {
+			node := ep.node.Name
+			zone := ep.node.Labels[v1.LabelTopologyZone]
```
The only thing that changes depending on `nodesHaveZones` is whether we expect the EndpointSlices in the PreferSameNode case to also have zone hints or _not_ also have zone hints. The actual expected results of the test (which clients talk to which servesr) don't change.

Though maybe I should just remove the EndpointSlice hint checking entirely... the unit and integration tests already check that, and we're supposed to be checking the end-to-end behavior here, not the low-level details. (If something fails we can dump the EndpointSlice for debugging.)

aojea commented:
is it in draft mode on purpose?

danwinship commented:
(a) didn't want someone accidentally merging it, (b) didn't want all CI jobs to run automatically when I pushed

aojea commented:
> (a) didn't want someone accidentally merging it, (b) didn't want all CI jobs to run automatically when I pushed

do we need an exception? the tests are great, we have some mechanical comments that are solvable for this release?

danwinship commented:
The enhancement is already listed as "Tracked for code freeze"... I think because this PR is already `accepted`? Anyway, yes, I'm going to finish this.

danwinship committed:
Abstract the logic of the TrafficDistribution test

Split the logic of creating the clients and the servers apart from the
logic of checking which clients connect to which servers. Add some
extra complexity to support additional use cases (like multiple
endpoints on the same node).

danwinship committed:
Split out the pieces of the TrafficDistribution test

danwinship commented:
/test pull-kubernetes-e2e-kind-multizone

danwinship commented:
/test pull-kubernetes-e2e-kind-multizone

danwinship commented:
/retest
/milestone v1.33

gauravkghildiyal commented:
/lgtm

Thanks!
